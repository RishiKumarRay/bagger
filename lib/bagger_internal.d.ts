import { JSONSchema7 } from 'json-schema';
import { Schema } from '@hapi/joi';

type BaggerSchemas = JSONSchema7 | Schema;
