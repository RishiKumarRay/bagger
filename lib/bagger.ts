export const x: 'foo' | 'bar' = 'foo';
