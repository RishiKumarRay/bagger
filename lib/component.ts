import { Schema } from '@hapi/joi';
import { SchemaObject, ReferenceObject } from 'openapi3-ts';
import { createSwaggerDefinition } from './utils/create_swagger_definition';

export type ComponentType = 'schemas' | 'securitySchemes';

export interface SchemaComponentObject {
  [schema: string]: SchemaObject | ReferenceObject;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function isReference(value: any): value is ReferenceObject {
  return value.$ref && typeof value.$ref === 'string';
}

export class BaggerSchemaComponent {
  private name: string;
  private schema: Schema | ReferenceObject;

  public constructor(name: string, schema: Schema) {
    this.name = name;
    this.schema = schema;
  }

  private compileValue(): SchemaObject | ReferenceObject {
    if (isReference(this.schema)) {
      return this.schema;
    } else {
      return createSwaggerDefinition(this.schema);
    }
  }

  public compile(): { name: string; schema: SchemaObject | ReferenceObject } {
    return {
      name: this.name,
      schema: this.compileValue()
    };
  }
}
