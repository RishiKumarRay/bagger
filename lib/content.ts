import { createSwaggerDefinition } from './utils/createSwaggerDefinition';
import { JSONSchema7 } from 'json-schema';
import { Schema } from '@hapi/joi';
import { SchemaObject, ContentObject } from 'openapi3-ts';

// There are no type definitions for json-schema-to-openapi-schema
// eslint-disable-next-line @typescript-eslint/no-var-requires
const toOpenApi = require('json-schema-to-openapi-schema');

interface InternalContent {
  mediaType: string;
  schema: SchemaObject;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function isJoi(schema: any): schema is Schema {
  return schema.isJoi;
}

export class Content {
  private internals: InternalContent;

  public constructor(mediaType: string, schema: JSONSchema7 | Schema) {
    if (isJoi(schema)) {
      this.internals = { mediaType, schema: createSwaggerDefinition(schema) };
    } else {
      this.internals = { mediaType, schema: toOpenApi(schema) };
    }
  }

  public compile(): ContentObject {
    const { mediaType, schema } = this.internals;
    return {
      [mediaType]: {
        schema
      }
    };
  }
}
