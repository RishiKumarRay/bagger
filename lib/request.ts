import { parse as validateContentType } from 'content-type';
import { BaggerResponse } from './response';
import { RequestBody, BaggerRequestBody } from './request_body';
import { PathItemObject } from 'openapi3-ts';

type Method = 'get' | 'post' | 'patch' | 'put' | 'delete' | 'options' | 'head' | 'trace' | 'connect';

interface RequestContext {
  path: string;
  summary?: string;
  description?: string;
  methods: Set<Method>;
  tags: Set<string>;
  security: Set<string>;
  produces: Set<string>;
  requestBody?: RequestBody | BaggerRequestBody;
  responses?: Set<BaggerResponse>;
}

interface CompiledMethodContent {
  summary?: string;
  description?: string;
  tags?: Set<string>;
  security?: Set<string>;
  produces?: Set<string>;
  requestBody?: RequestBody | BaggerRequestBody;
  responses?: Set<BaggerResponse>;
}

export interface CompiledRequest {
  [path: string]: {
    [method in Method]: CompiledMethodContent;
  };
}

export class BaggerRequest {
  /*private _path: string = '';
  private pathContext: PathItemObject;*/
  private context: RequestContext = {
    // TODO: Use open api types
    path: '', // TODO: should not have default value
    methods: new Set<Method>(),
    tags: new Set<string>(),
    security: new Set<string>(),
    produces: new Set<string>()
  };
  public readonly isBagger = true;

  public path(path: string): BaggerRequest {
    if (!this.context.path) {
      this.context.path = path;
    } else {
      console.error('path set multiple times');
    }
    return this;
  }

  public method(method: Method): BaggerRequest {
    if (!this.context.methods) {
      this.context.methods = new Set<Method>();
    }
    this.context.methods.add(method);
    return this;
  }

  public tag(tag: string): BaggerRequest {
    if (!this.context.tags) {
      this.context.tags = new Set<string>();
    }
    this.context.tags.add(tag);
    return this;
  }

  public security(scheme: string): BaggerRequest {
    if (!this.context.security) {
      this.context.security = new Set<string>();
    }
    this.context.security.add(scheme);
    return this;
  }

  public produces(contentType: string): BaggerRequest {
    if (!this.context.produces) {
      this.context.produces = new Set<string>();
    }
    try {
      validateContentType(contentType);
      this.context.produces.add(contentType);
    } catch (err) {
      console.error('invalid content type', err);
    }
    return this;
  }

  public body(requestBody: RequestBody | BaggerRequestBody): BaggerRequest {
    if (!this.context.requestBody) {
      this.context.requestBody = requestBody;
    } else {
      console.error('request body set multiple times');
    }
    return this;
  }

  public response(response: BaggerResponse): BaggerRequest {
    if (!this.context.responses) {
      this.context.responses = new Set<BaggerResponse>();
    }
    this.context.responses.add(response);
    return this;
  }

  public compile(): CompiledRequest {
    const obj: PathItemObject = {};
    obj[this.context.path] = {};
    /* eslint-disable */
    const v: any = this.context as CompiledMethodContent;
    this.context.methods.forEach((value: Method, _: Method, _set: Set<Method>) => {
      const newVal = Object.keys(v).reduce(
        (prev, curr) => {
          let val = v[curr];
          if (typeof v[curr].compile === 'function') {
            val = v[curr].compile();
          } else if (v[curr] instanceof Set) {
            val = Array.from(v[curr]);
          }
          prev[curr] = val;
          return prev;
        },
        {} as any
      );
      obj[this.context.path][value] = newVal;
    });
    /* eslint-enable */
    return obj;
  }
}
