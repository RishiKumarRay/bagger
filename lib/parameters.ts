import { ParameterObject, ParameterStyle, SchemaObject, ReferenceObject, ExampleObject } from 'openapi3-ts';
import { Content } from './content';

type ExamplesObject = {
  [param: string]: ExampleObject | ReferenceObject;
};

export class BaggerParameter {
  private settings: ParameterObject = {
    in: 'path',
    name: 'path'
  };

  public query(): BaggerParameter {
    this.settings.in = 'query';
    return this;
  }

  public path(): BaggerParameter {
    this.settings.in = 'path';
    return this;
  }

  public cookie(): BaggerParameter {
    this.settings.in = 'cookie';
    return this;
  }

  public header(): BaggerParameter {
    this.settings.in = 'header';
    return this;
  }

  public name(name: string): BaggerParameter {
    this.settings.name = name;
    return this;
  }

  public description(description: string): BaggerParameter {
    this.settings.description = description;
    return this;
  }

  public required(required: boolean): BaggerParameter {
    this.settings.required = required;
    return this;
  }

  public deprecated(deprecated: boolean): BaggerParameter {
    this.settings.deprecated = deprecated;
    return this;
  }

  public allowEmptyValue(allowEmptyValue: boolean): BaggerParameter {
    this.settings.allowEmptyValue = allowEmptyValue;
    return this;
  }

  public style(style: ParameterStyle): BaggerParameter {
    this.settings.style = style;
    return this;
  }

  public explode(explode: boolean): BaggerParameter {
    this.settings.explode = explode;
    return this;
  }

  public allowReserved(allowReserved: boolean): BaggerParameter {
    this.settings.allowReserved = allowReserved;
    return this;
  }

  public schema(schema: SchemaObject | ReferenceObject): BaggerParameter {
    this.settings.schema = schema;
    return this;
  }

  public examples(examples: ExamplesObject): BaggerParameter {
    this.settings.examples = examples;
    return this;
  }

  public content(content: Content): BaggerParameter {
    this.settings.content = content.compile();
    return this;
  }

  public compile(): ParameterObject {
    return this.settings;
  }
}
