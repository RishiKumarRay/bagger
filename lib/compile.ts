import {
  OpenAPIObject,
  InfoObject,
  ServerObject,
  PathsObject,
  SecurityRequirementObject,
  ExternalDocumentationObject,
  ComponentsObject
} from 'openapi3-ts';
import { BaggerRequest } from './request';
import { cleanObject } from './utils/clean_object';
import { validateSchema } from './utils/validate_schema';

interface SwaggerConfiguration {
  info: InfoObject;
  servers?: ServerObject[];
  security?: SecurityRequirementObject[];
  externalDocs?: ExternalDocumentationObject;
}

function compileRequests(requests: BaggerRequest[]): PathsObject {
  const compiledRequests = requests.map(request => request.compile());

  const mergedRequests = compiledRequests.reduce((paths, request) => {
    const requestObject = Object.entries(request)[0];
    const [path, definition] = requestObject;

    paths[path] = {
      ...definition,
      ...paths[path]
    };

    return paths;
  }, {});

  return mergedRequests;
}

/**
 * Compiles all requests into a Swagger definition.
 * @param configuration General configuration settings describing the API.
 * @param requests All requests that together describe the API.
 * @param components Any additional components that are externally defined.
 */
export function compile(
  configuration: SwaggerConfiguration,
  requests: BaggerRequest[],
  components?: ComponentsObject
): OpenAPIObject {
  const paths = compileRequests(requests);
  const swaggerDefinition = cleanObject({
    ...configuration,
    openapi: '3.0.0',
    paths,
    components
  });

  const errors = validateSchema(swaggerDefinition);

  if (errors.length > 0) {
    throw new Error(JSON.stringify(errors));
  }

  return swaggerDefinition;
}
