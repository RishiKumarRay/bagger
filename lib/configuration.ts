import { BaggerRequest } from './request';
import {
  ExternalDocumentationObject,
  SecurityRequirementObject,
  ServerObject,
  InfoObject,
  OpenAPIObject,
  PathsObject
} from 'openapi3-ts';
import { cleanObject } from './utils/clean_object';

interface SwaggerConfiguration {
  info: InfoObject;
  servers?: ServerObject[];
  security?: SecurityRequirementObject[];
  externalDocs?: ExternalDocumentationObject;
}

export class BaggerConfiguration {
  private internalConfiguration: BaggerConfigurationInternal;

  public constructor(internalConfiguration: BaggerConfigurationInternal) {
    this.internalConfiguration = internalConfiguration;
  }

  public info(info: InfoObject): BaggerConfiguration {
    this.internalConfiguration.setInfo(info);
    return this;
  }

  public addServer(server: ServerObject): BaggerConfiguration {
    this.internalConfiguration.addServer(server);
    return this;
  }

  public addSecurity(security: SecurityRequirementObject): BaggerConfiguration {
    this.internalConfiguration.addSecurity(security);
    return this;
  }

  public externalDocs(externalDocs: ExternalDocumentationObject): BaggerConfiguration {
    this.internalConfiguration.setExternalDocs(externalDocs);
    return this;
  }
}

export class BaggerConfigurationInternal {
  private requests: BaggerRequest[] = [];
  private configuration: SwaggerConfiguration = {
    info: {
      title: '',
      version: ''
    }
  };

  public addRequest(request: BaggerRequest): void {
    this.requests.push(request);
  }

  public setInfo(info: InfoObject): void {
    this.configuration.info = info;
  }

  public addServer(server: ServerObject): void {
    if (!this.configuration.servers) {
      this.configuration.servers = [];
    }
    this.configuration.servers.push(server);
  }

  public addSecurity(security: SecurityRequirementObject): void {
    if (!this.configuration.security) {
      this.configuration.security = [];
    }
    this.configuration.security.push(security);
  }

  public setExternalDocs(externalDocs: ExternalDocumentationObject): void {
    this.configuration.externalDocs = externalDocs;
  }

  private mergeRequests(): PathsObject {
    const mergedRequests = this.requests
      .map(request => request.compile())
      .reduce((paths, request) => {
        const requestObject = Object.entries(request)[0];
        const [path, definition] = requestObject;

        paths[path] = {
          ...definition,
          ...paths[path]
        };

        return paths;
      }, {});

    return mergedRequests;
  }

  public compile(): OpenAPIObject {
    const paths = this.mergeRequests();
    return cleanObject({
      ...this.configuration,
      openapi: '3.0.0',
      paths
    });
  }
}
