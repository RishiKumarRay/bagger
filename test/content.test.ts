import { Content } from '../lib/content';

describe('Content objects', () => {
  test('Simple content object can be created and compiled', () => {
    const content = new Content();
    content.add('application/json', { type: 'string', examples: ['foo'] });
    expect(content.compile()).toMatchSnapshot();
  });
});
