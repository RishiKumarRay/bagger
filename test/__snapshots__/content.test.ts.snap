// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Content objects Simple content object can be created and compiled 1`] = `
Object {
  "application/json": Object {
    "schema": Object {
      "examples": Array [
        "foo",
      ],
      "type": "string",
    },
  },
}
`;
